shopfloor/services/delivery.py

```diff
@@ -54,21 +135,200 @@ def scan_deliver(self, barcode):
         When all the available move lines of the stock picking are done, the
         stock picking is set to done.
 
+        The ``picking_id`` parameter is used to be stateless: if the client
+        sends a wrong barcode, it allows to stay on the last picking with
+        updated data (and we really want to refresh data because several
+        users may work on the same transfer).
+
         Transitions:
-        * deliver: always return here with the data for the last touched picking
-        or no picking if the picking has been set to done
+        * deliver: always return here with the data for the last touched
+        picking or no picking if the picking has been set to done
         """
-        return self._response()
+        search = self.actions_for("search")
+        picking = search.picking_from_scan(barcode)
+        if picking:
+            response = self._check_picking_status(picking)
+            if response:
+                return response
+            return self._response_for_deliver(picking=picking)
+
+        # We should have only a picking_id because the client was working
+        # on it already, so no need to validate the picking type
+        if picking_id:
+            picking = self.env["stock.picking"].browse(picking_id).exists()
+
+        package = search.package_from_scan(barcode)
+        if package:
+            return self._deliver_package(picking, package)
+
+        product = search.product_from_scan(barcode)
+        if product:
+            return self._deliver_product(picking, product)
+
+        lot = search.lot_from_scan(barcode)
+        if lot:
+            return self._deliver_lot(picking, lot)
+
```

Comment:
you should call `_action_picking_done` here.
Actually `set_qty_done_pack` and `set_qty_done_line` are never called because everything happens on scan.
@guewen ?

Reply:
well, you can call `set_qty_done_line` and `set_qty_done_pack` from `scan_deliver` instead.

Reply:
It should be called every time we call `_set_lines_done`, so in `_deliver_package` too. Or... called at the end of `_set_lines_done`.